### DISCUSSION TITLE
Complex Permissions: Hybrid ABAC -> RBAC

### DISCUSSION THREAD

**Discussion by kitsunemax** [blitz-js/blitz#2116] [Q&A]
We have an in-app permissions setup, which is familiar to all of us, but does not seem to be as easy to config as one would hope.

Basically, it is permissions like you are used to with google drive, airtable, slack, your PM/task tools, etc .... and that is that you give someone access to a certain type of asset, with a certain permissions level for that asset.
Additional, this permission is usually inherited to child assets in the hierarchy with the same permissions, unless a child has their permissions over-ridden as private, etc.

The access to these assets is based off the asset "type" being shared with other users, within a hierarchy, and for certain permissions (ie crud, etc) for each asset that is shared with the other user.

To us, this seems to be hybrid ABAC -> RBAC ... as in, the user gets permissions to a certain assetID and type, and for that assetID and type, they are granted certain roles, for that asset type.

---
For us, we have a number of assets within a hierarchy, such as:
- organisation {one:many} child-organisation
- project {one:many} child-project
- dashboard {one:many} tile {many:many} data-connector, etc ....

And for each asset type, we have certain roles. such as:
- owner, manager, collaborator, guest-edit, guest-view, public, etc

---
**Example**
> This means that the OWNER of dashboardID9akje86874, within projectID987365, which is within orgID362,  wishes to share that dashboard with userIDkjshu8r6783 and give them the permissions of a guest-view for ONLY dashboardID9akje86874.

userIDkjshu8r6783 can only view the dashboard, they are unable to see detail of the org or the project, unless it is displayed on the dashboard.  
*(like when someone shares a file or folder with you from google drive, you can see what is within that asset that is shared with you, but not the parent structure/assets).*

That same OWNER has permissions to add someone else to projectID987365 with collaborator permissions on the project.  They would then be able to see the project details and all the dashboards in the project.

---
We are currently investigating the following libraries, which seem to solidly support hybrid ABAC/RBAC permissions scenario:
* https://casbin.org/
* https://www.osohq.com/

Has anyone looked at these, and any thoughts on which may/may not work best with BlitzJS ?

Thanks in advance for any thoughts, advice or assistance on how to approach.

Max

---
ps: have looked at blitzguard and other threads, but have not seen anything that is quite able to do this, and we've been working around it for a while.  In [blitzguard can see resource TYPES](https://ntgussoni.github.io/blitz-guard/docs/resources#extending-the-resources), then roles .... but not individual assetIDs, then type, then roles.
pps: We also tried approaching in schema but only takes so far, and is easy to have this in the prisma2 structure (and prisma is great to help with this once you work it out), but the per asset, per type, then roles on top of that being managed in a scalable way, then retrievable in the UI for a userID, or an assetID, and for sharing on individual asset in a hierarchy needs a bit more heavy lifting (and my be more than what we can spin up ourselves with any surety we haven't stuffed it up).   
Would love to work this out, as will help us a HUGE deal to move forward, as permissions is where we are blocked at the moment  :frowning: 

---
Refs:
- https://github.com/ntgussoni/blitz-guard
- https://github.com/blitz-js/blitz/discussions/1978
- https://blitzjs.com/docs/authorization
- https://github.com/blitz-js/blitz/discussions/756

**Comment by flybayer** [comment]
Hey @MaxLynam unfortunately I don't think anyone has tackled this yet. But I'm more than happy to help you work through this, get on brainstorming calls, etc. We definitely need a good solution for this.

**Reply by kitsunemax** [reply]
thanks @flybayer 🤩  

@sirkismatt and I are in Melbourne Australian (GMT+11) and New Zealand (GMT+9) ... along with most of our contributors.

Usually from the US, your late afternoon is our early (but acceptable) morning, is you wanted to hook us up for a quick call (max@forhumanity.org.au) -> or if there's another initial step (docs, flow, etc??) please let me know next steps.

Thanks again .... this is a huge thing for us, and many apps - so would be super-cool to get it sorted - thank you  🙏🏽

**Comment by kitsunemax** [comment]
Follow up from this approach like (with user > org > assets):
https://blog.bullettrain.co/teams-should-be-an-mvp-feature/

BUT ... for having the right info stored against the userID by the asset type and IDs, perhaps something like permissions (membership) table with:
```
UserID {
orgOwn: id1id,2,id3;
orgManage: id1id,2,id3;
orgCollab: id1id,2,id3;
etc
projOwn: 1id,2,id3;
projManage: id1id,2,id3:;
projCollab: id1id,2,id3;
etc
}
```
--or--
```
userID {
  {org
Own: id1id,2,id3;
Manage: id1id,2,id3;
Collab: id1id,2,id3;
etc
   }
   {proj
Own: 1id,2,id3;
Manage: id1id,2,id3;
Collab: id1id,2,id3;
etc
   }
}
```
... will have a play and see what we may come up with.
